From deb8dcad67418210d5cc2104a97cb8cb9fd66e5e Mon Sep 17 00:00:00 2001
From: Laszlo Agocs <laszlo.agocs@qt.io>
Date: Wed, 5 Aug 2026 15:53:12 +0200
Subject: [PATCH] rhi: d3d: Do derefAdapter while the LUID is still valid

Calling the function after adapterLuid is nulled out is clearly
not ideal.

Task-number: QTBUG-148834
Pick-to: 6.12 6.11 6.8
Change-Id: I16e3118fe700ec628e3da5557d101092dff03f13
Reviewed-by: Andy Nichols <andy.nichols@qt.io>

Backport: Adjusted the D3D12 context to Qt 6.11.2 while preserving the upstream release ordering.

---
diff --git a/src/gui/rhi/qrhid3d11.cpp b/src/gui/rhi/qrhid3d11.cpp
index a252ff2..d9b25ca 100644
--- a/src/gui/rhi/qrhid3d11.cpp
+++ b/src/gui/rhi/qrhid3d11.cpp
@@ -461,10 +461,10 @@ void QRhiD3D11::destroy()
         dxgiFactory = nullptr;
     }
 
+    QDxgiVSyncService::instance()->derefAdapter(adapterLuid);
+
     importedDeviceAndContext = false;
     adapterLuid = {};
-
-    QDxgiVSyncService::instance()->derefAdapter(adapterLuid);
 }
 
 void QRhiD3D11::reportLiveObjects(ID3D11Device *device)
diff --git a/src/gui/rhi/qrhid3d12.cpp b/src/gui/rhi/qrhid3d12.cpp
index 2ded95f..b618ed2 100644
--- a/src/gui/rhi/qrhid3d12.cpp
+++ b/src/gui/rhi/qrhid3d12.cpp
@@ -621,11 +621,11 @@ void QRhiD3D12::destroy()
         dxgiFactory = nullptr;
     }
 
+    QDxgiVSyncService::instance()->derefAdapter(adapterLuid);
+
     adapterLuid = {};
     importedDevice = false;
     importedCommandQueue = false;
-
-    QDxgiVSyncService::instance()->derefAdapter(adapterLuid);
 }
 
 QRhi::AdapterList QRhiD3D12::enumerateAdaptersBeforeCreate(QRhiNativeHandles *nativeHandles) const
